File: features/mongo/step_definitions/wire_steps.rb
# step definitions for wire protocol
Given /^my request will have an ID of (\d+)$/ do |id|
  @id = id.to_i
end

Given /^I am using the collection "(.*?)"$/ do |coll_name|
  @coll_name = coll_name
end

Given /^I am (generating an? \S+ message)$/ do |op_class|
  @msg_class = op_class
end

Given /^I am selecting fields to update by (the document .*)$/ do |doc|
  @update_select_doc = doc
end

Given /^I am selecting fields to return by (the document .*)$/ do |doc|
  @return_select_doc = doc
end

Given /^I am updating by (the document .*)$/ do |doc|
  @update_doc = doc
end

Given /^I am querying by (the document .*)$/ do |doc|
  @query_doc = doc
end

Given /^I am selecting documents to delete by (the document .*)$/ do |doc|
  @delete_doc = doc
end

Given /^I am requesting results for the cursor with id (\d+)$/ do |cur_id|
  @cursor_to_get_more = cur_id.to_i
end

Given /^I am inserting the documents:$/ do |table|
  # table is a Cucumber::Ast::Table
  @docs_to_insert = table.rows.map do |row|
    JSON[row.first]
  end
end

Given /^I am deleting the cursors with ids:$/ do |table|
  # table is a Cucumber::Ast::Table
  @cursors_to_delete = table.rows.map do |row|
    row.first.to_i
  end
end

# non-binary options
Given /^I am skipping (\d+) results$/ do |num|
  @num_skip_results = num.to_i
end

Given /^I am returning (\d+) results$/ do |num|
  @num_return_results = num.to_i
end

Given /^MongoDB has responded with the OP_REPLY message (\S+)$/ do |wire|
  @wire = [wire].pack('H*')
end

# actually, message header generation. for ensuring randomness of IDs, headers are all we need
Given /^I have generated a message$/ do
  @header_1 = Mongo::Wire::MessageHeader.new
end

# binary flags below
# the weird or-ing stuff has to do with needing to tolerate the difference
# between human-generated and table-generated scenarios
# (the latter may have an extra space)
Given /^(I am (?:not | |)doing an upsert)$/ do |bool|
  @upsert = bool
end

Given /^(I am (?:not | |)doing a multi update)$/ do |bool|
  @multi_update = bool
end

Given /^(I am (?:not | |)doing a continue on error)$/ do |bool|
  @continue_on_error = bool
end

Given /^(I am (?:not | |)doing a tailable cursor query)$/ do |bool|
  @tailable_cursor = bool
end

Given /^(I am (?:not | |)permitting querying of a replica slave)$/ do |bool|
  @slave_ok = bool
end

Given /^(I am (?:not | |)permitting idle cursors to persist)$/ do |bool|
  @no_cursor_timeout = bool
end

Given /^(I am (?:not | |)permitting cursors to block and wait for more data)$/ do |bool|
  @await_data = bool
end

Given /^(I am (?:not | |)pulling all queried data at once)$/ do |bool|
  @exhaust = bool
end

Given /^(I am (?:not | |)permitting partial results if a shard is down)$/ do |bool|
  @partial = bool
end

Given /^(I am (?:not | |)permitting removal of multiple documents)$/ do |bool|
  @single_remove = (not bool)
end


When /^I generate the wire protocol message for this request$/ do
  @msg = @msg_class.new
  @msg.get_header.request_id(@id)

  # hack to make the case statement comparisons work
  case [@msg_class]
  when [Mongo::Wire::RequestMessage::Update]
    flags = Mongo::Wire::RequestMessage::Update::RequestFlags.new
    flags.upsert(@upsert).multi_update(@multi_update)
    @msg.flags(flags)
        .full_collection_name(@coll_name).selector(@update_select_doc).update(@update_doc)

  when [Mongo::Wire::RequestMessage::Insert]
    flags = Mongo::Wire::RequestMessage::Insert::RequestFlags.new
    flags.continue_on_error(@continue_on_error)
    @msg.flags(flags)
        .full_collection_name(@coll_name).documents(@docs_to_insert)

  when [Mongo::Wire::RequestMessage::Query]
    flags = Mongo::Wire::RequestMessage::Query::RequestFlags.new
    flags.tailable_cursor(@tailable_cursor).slave_ok(@slave_ok)
         .no_cursor_timeout(@no_cursor_timeout).await_data(@await_data)
         .exhaust(@exhaust).partial(@partial)
    @msg.flags(flags)
        .full_collection_name(@coll_name).number_to_skip(@num_skip_results)
        .number_to_return(@num_return_results).query(@query_doc)
        .return_field_selector(@return_select_doc)

  when [Mongo::Wire::RequestMessage::GetMore]
    @msg.full_collection_name(@coll_name).number_to_return(@num_return_results)
        .cursor_id(@cursor_to_get_more)

  when [Mongo::Wire::RequestMessage::Delete]
    flags = Mongo::Wire::RequestMessage::Delete::RequestFlags.new
    flags.single_remove(@single_remove)
    @msg.flags(flags)
        .full_collection_name(@coll_name).selector(@delete_doc)

  when [Mongo::Wire::RequestMessage::KillCursors]
    @msg.cursor_ids(@cursors_to_delete)

  end

  @wire = @msg.to_wire
end

When /^I parse the message$/ do
  @msg = Mongo::Wire::ResponseMessage::Reply.new(@wire)
end

When /^I generate another message$/ do
  @header_2 = Mongo::Wire::MessageHeader.new
end


Then /^the generated message should match (\S+) ?$/ do |msg|
  @wire.should == [msg].pack('H*')
end

Then /^I should learn that the message has identifier (\d+)$/ do |msg_id|
  @msg.header.get_request_id.should == msg_id.to_i
end

Then /^I should learn that the message is in response to the request with identifier (\d+)$/ do |resp_id|
  @msg.header.get_response_to.should == resp_id.to_i
end

Then /^the opcode should correspond to OP_REPLY$/ do
  #  TODO: make it a constant
  @msg.header.get_message_class.should == Mongo::Wire::ResponseMessage::Reply
end

Then /^I should learn that the cursor ID to get more results is (\d+)$/ do |cur_id|
  @msg.cursor_id.should == cur_id.to_i
end

Then /^I should learn that (\d+) documents are being returned$/ do |num_docs|
  @msg.number_returned.should == num_docs.to_i
end

Then /^I should learn that this reply is starting from (\d+) results into the cursor$/ do |start_from|
  @msg.starting_from.should == start_from.to_i
end

Then /^I should learn that the requested cursor( is (?:not | |))found$/ do |bool|
  @msg.flags.cursor_not_found.should == (not bool)
end
  
Then /^I should learn that the query( is (?:not | |))successful$/ do |bool|
  @msg.flags.query_failure.should == (not bool)
end

Then /^I should learn that the sharding configuration( is (?:not | |))outdated$/ do |bool|
  @msg.flags.shard_config_stale.should == bool
end

Then /^I should learn that the server( is (?:not | |))able to support the await_data query parameter$/ do |bool|
  @msg.flags.await_capable.should == bool
end


Then /^the message should contain the documents (.*)$/ do |docs_str|
  puts docs_str
  docs = JSON[docs_str]
  @msg.documents.should == docs
end


Then /^the two messages should not have the same request ID$/ do
  @header_1.get_request_id.should_not == @header_2.get_request_id
end

File: pymongo-test/features/steps/wire_steps.py
from behave import *
import pymongo
import bson
import StringIO
import struct
import re
import json
from wire_transforms import *
from utility import *
from nose.tools import *

# TODO - figure out which str() and strip() calls are actually necessary
# make it use regex, not the less powerful default parser
step_matcher('re')

# step definitions for wire protocol
@given('^my request will have an ID of (\d+)$')
def my_request_will_have_id(context, id):
  context.id = int(id)

@given('^I am using the collection "(.*)"$')
def i_am_using_collection(context, name):
  context.coll_name = str(name)

@step_tr(given, '^I am (generating an? \S+ message)$')
def i_am_generating_message(context, op_type):
  # transform gives us a pair: (name, function)
  # TODO this transform may no longer be necessary
  context.msg_type = op_type[0]
  context.msg_function = op_type[1]

@step_tr(given, '^I am performing the update specified by (the document .*)$')
def i_am_performing_update_specified(context, doc):
  context.update_spec_doc = doc

@step_tr(given, '^I am selecting fields to return by (the document .*)$')
def i_am_selecting_return_field(context, doc):
  context.return_select_doc = doc

@step_tr(given, '^I am updating(?: the)? documents? matching (the document .*)$')
def i_am_updating_documents(context, doc):
  context.update_doc = doc

@step_tr(given, '^I am querying for documents matching (the document .*)$')
def i_am_querying_documents(context, doc):
  context.query_doc = doc

@step_tr(given, '^I am deleting documents matching (the document .*)$')
def i_am_deleting_documents(context, doc):
  context.delete_doc = doc

@given('^I am requesting results for the cursor with id (\d+)$')
def i_am_requesting_cursor_results(context, cur_id):
  context.cursor_to_get_more = int(cur_id)

@step_tr(given, '^I am inserting (the document .*)$')
def i_am_inserting_document(context, doc):
  context.docs_to_insert = [doc]

@given('^I am inserting the documents$')
def i_am_inserting_documents(context):
  docs = []
  for row in context.table:
    print "item!: ", row[0]
    doc = json.loads(row[0].strip())
    docs.append(doc)

  context.docs_to_insert = docs

@given('^I am deleting the cursors with ids$')
def i_am_deleting_cursors(context):
  cursors = []
  for row in context.table:
    cursors.append(int(row[0]))

  context.cursors_to_delete = cursors

# non-binary options
@given('^I am skipping (\d+) results$')
def i_am_skipping_results(context, num):
  context.skip_num = int(num)

@given('^I am returning (\d+) results$')
def i_am_returning_results(context, num):
  context.return_num = int(num)

@given('^MongoDB has responded with the OP_REEPLY message (\S+)$')
def mongodb_responded_with_reply(context, wire):
  context.wire = wire.encode('hex')

@given('^I have generated a message$')
def i_generated_message(context):
  # noop
  return

# binary flags

@step_tr(given, '^(I am (?:not | |)doing an upsert)$')
def i_am_doing_upsert(context, yesno):
  context.upsert = yesno

@step_tr(given, '^(I am (?:not | |)doing a multi update)$')
def i_am_doing_multi_update(context, yesno):
  context.multi_update = yesno

@step_tr(given, '^(I am (?:not | |)doing a continue on error)$')
def i_am_doing_continue_error(context, yesno):
  context.continue_on_error = yesno

@step_tr(given, '^(I am (?:not | |)doing a tailable cursor query)$')
def i_am_doing_tailable_cursor(context, yesno):
  context.tailable_cursor = yesno

@step_tr(given, '^(I am (?:not | |)permitting querying of a replica slave)$')
def i_am_permitting_replica_slave_query(context, yesno):
  context.slave_ok = yesno

@step_tr(given, '^(I am (?:not | |)permitting idle cursors to persist)$')
def i_am_permitting_idle_cursors(context, yesno):
  context.no_cursor_timeout = yesno

@step_tr(given, '^(I am (?:not | |)permitting cursors to block and wait for more data)$')
def i_am_permitting_cursors_to_await(context, yesno):
  context.await_data = yesno

@step_tr(given, '^(I am (?:not | |)pulling all queried data at once)$')
def i_am_pulling_all_data_at_once(context, yesno):
  context.exhaust = yesno

@step_tr(given, '^(I am (?:not | |)permitting partial results if a shard is down)$')
def i_am_permitting_partial_results(context, yesno):
  context.partial = yesno

@step_tr(given, '^(I am (?:not | |)permitting removal of multiple documents)$')
def i_am_permitting_multiple_documents_remove(context, yesno):
  context.single_remove = (not yesno)

# actually generating the message!
# TODO - UUID subtype will change
@when('^I generate the wire protocol message for this request$')
def i_generate_wire_protocol_message(context):
  typ = context.msg_type
  if   typ == 'OP_UPDATE':
    msg = pymongo.message.update(
      context.coll_name, context.upsert, context.multi_update,
      context.update_spec_doc, context.update_doc,
      False, [], False, bson.binary.OLD_UUID_SUBTYPE
    )

  elif typ == 'OP_INSERT':
    print "about to insert"
    print context.docs_to_insert
    msg = pymongo.message.insert(
      context.coll_name, context.docs_to_insert, 
      False, False, [], context.continue_on_error, bson.binary.OLD_UUID_SUBTYPE
    )

  elif typ == 'OP_QUERY':
    # we have to put together our query options by hand.
    flags = 0
    # bit 0 reserved
    flags |= (0b00000010 if context.tailable_cursor else 0)
    flags |= (0b00000100 if context.slave_ok else 0)
    # driver does not set bit 3
    flags |= (0b00010000 if context.no_cursor_timeout else 0)
    flags |= (0b00100000 if context.await_data else 0)
    flags |= (0b01000000 if context.exhaust else 0)
    flags |= (0b10000000 if context.partial else 0)

    msg = pymongo.message.query(
      flags, context.coll_name, context.skip_num, context.return_num,
      context.query_doc, context.return_select_doc,
      bson.binary.OLD_UUID_SUBTYPE
    )

  elif typ == 'OP_GET_MORE':
    msg = pymongo.message.get_more(
      context.coll_name, context.return_num, context.cursor_to_get_more
    )

  elif typ == 'OP_DELETE':
    msg = pymongo.message.delete(
      context.coll_name, context.delete_doc,
      False, [], bson.binary.OLD_UUID_SUBTYPE
    )

  elif typ == 'OP_KILL_CURSORS':
    msg = pymongo.message.kill_cursors(context.cursors_to_delete)

  else:
    msg = ""
    print "invalid message type ", typ
    assert(False) # if we get here something is wrong with our scenario or steps

  context.wire = msg

@when('^I parse the message$')
def i_parse_message(context):
  context.response_props = pymongo.helpers._unpack_response(context.msg)

@then('^the generated message should match (\S+) ?$')
def generated_message_should_match(context, msg):
  generated_message = context.wire[1] # we get back a tuple apparently
  # remove request IDs, since they won't reliably be the same
  msg = pack_hex(msg)
  assert_equal_irrespective_of_ids( generated_message, msg )

# @then('^I should learn that the message has identifier (\d+)')
# def i_should_learn_message_id(context, id):
#   assert_equal ( context.response_props[""], int(id) )

# @then('^I should learn that the message is in response to the request with identifier (\d+)$')
# def i_should_learn_message_id(context, id):
#   assert_equal

# @then('^the opcode should correspond to OP_REPLY$')

@then('^I should learn that the cursor ID to get more results is (\d+)$')
def i_should_learn_cursor_id(context, cur_id):
  assert_equal( context.response_props["cursor_id"], int(cur_id) )

@then('^I should learn that (\d+) documents are being returned$')
def i_should_learn_num_documents_returned(context, num_docs):
  assert_equal( context.response_props["number_returned"], int(num_docs) )

@then('^I should learn that this reply is starting from (\d+) results into the cursor$')
def i_should_learn_reply_start_from(context, start_from):
  assert_equal( context.response_props["starting_from"], int(start_from) )

#
# error-checking code doesn't seem to translate well to python
#

@then('^the message should contain the documents (.*)$')
def message_should_contain_documents(context, docs_str):
  docs = json.loads(docs_str.strip())
  assert_equal( context.documents, docs )

# @then('^the two messages should not have the same request ID$')